refactor: Mark functions async in `aws/customResources` and `aws/deploy`

ifitzsimmons opened the pull request:
Refactored Promise-returning functions for:
* `lib/plugin/aws/customResources`
* `lib/plugin/aws/deploy`

I had to refactor a few tests after tagging functions with `async` keyword. I didn't change the functionality of the test, just the syntax.
<!-- ⚠️⚠️ Acknowledge ALL below remarks -->
<!-- ⚠️⚠️ PR will not be processed if it doesn't meet outlined criteria -->

<!-- ⚠️⚠️ Do not propose PR's without prior agreement on a solution in the corresponding issue -->
<!-- ⚠️⚠️ Only documentation updates and obvious bug fixes are welcome without it -->

<!--
⚠️⚠️ Ensure to follow code style guidelines
https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md#code-style
-->

<!--
⚠️⚠️ Ensure to cover changes with tests written according to test guidelines
https://github.com/serverless/serverless/blob/master/test/README.md
-->

<!-- ⚠️⚠️ Ensure that support for Node.js v10 is maintained. -->

<!--
⚠️⚠️ Ensure that the proposed change passes CI. Confirm that by running the following scripts:
• npm run prettier-check
• npm run lint
• npm test
-->

<!--
⚠️⚠️ If proposed change touches integration with AWS services, confirm integration tests pass:
https://github.com/serverless/serverless/blob/master/test/README.md#aws-integration-tests
-->

<!-- ⚠️⚠️ After your PR is submitted, review the final CI status and address eventual failure -->

<!-- ⚠️⚠️ Answer below questions -->

<!--
Q1: Provide a link to the corresponding issue

• If PR *partially* addresses issue, ensure to rename "Closes" to "Addresses" ("Closes" term will automatically close an issue on PR merge)
• If it's a documentation update or obvious bug fix that has no corresponding issue, replace this line with a short description of made changes
-->

Addresses: #8368

pgrzesik commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -53,7 +53,7 @@ module.exports = {
       typeof this.provider.serverless.service.provider.stackName === 'string'
     ) {
-      return `${this.provider.serverless.service.provider.stackName}`;
+      return this.provider.serverless.service.provider.stackName;
```
good catch :+1:

pgrzesik commented on test/unit/lib/plugins/aws/deploy/lib/extendedValidate.test.js:
```diff
@@ -90,27 +90,25 @@ describe('extendedValidate', () => {
-        expect(fileExistsSyncStub.calledOnce).to.equal(true);
-        expect(readFileSyncStub.calledOnce).to.equal(true);
-      });
+      await awsDeploy.extendedValidate();
```
Does it have to be `await`-ed here? According to the changes to `extendedValidate`, that function is no longer async

ifitzsimmons replied:
Ah, good catch. I'll fix all of those `await`'s.

pgrzesik commented on test/unit/lib/plugins/aws/deploy/lib/extendedValidate.test.js:
```diff
@@ -90,27 +90,25 @@ describe('extendedValidate', () => {
-        expect(fileExistsSyncStub.calledOnce).to.equal(true);
-        expect(readFileSyncStub.calledOnce).to.equal(true);
-      });
+      await awsDeploy.extendedValidate();
```
Same as above

pgrzesik commented on test/unit/lib/plugins/aws/deploy/lib/extendedValidate.test.js:
```diff
@@ -156,18 +153,18 @@ describe('extendedValidate', () => {
-        delete awsDeploy.serverless.service.package.artifact;
-      });
+
+      await awsDeploy.extendedValidate();
```
Same here

pgrzesik commented on test/unit/lib/plugins/aws/deploy/lib/extendedValidate.test.js:
```diff
@@ -185,13 +182,13 @@ describe('extendedValidate', () => {
-        ].join('');
-        expect(awsDeploy.serverless.cli.log.firstCall.calledWithExactly(msg)).to.be.equal(true);
-      });
+      await awsDeploy.extendedValidate();
```
same here

pgrzesik reviewed (changes requested):
Thank you @ifitzsimmons :bow: Overall it looks good, I've had a few minor comments, but nothing major :+1:

ifitzsimmons commented:
@pgrzesik, I'm having a little trouble merging master back in to my branch. It looks like once I merge it, there are some 
strange errors that don't have much to do with my branch. At the very last, there weren't any manual merges required for any of those files.

![image](https://user-images.githubusercontent.com/43529054/103598103-c8dc2700-4ecf-11eb-911c-59a17d0b6945.png)

pgrzesik commented:
Hey @ifitzsimmons, sorry for the trouble. Yesterday we had some upgrades to internal dependencies which are most likely causing issues for you - reinstalling all deps locally after you merge master back or rebase your branch on top of it should do the trick. Also, we switched to `v2` of `prettier`, so there might be some minor conflicts for you to resolve.

ifitzsimmons commented:
Okay cool, I wasn't sure how you felt about a rebase during the review process but I'll do that today.

ifitzsimmons commented:
@pgrzesik I got dinged for a dip in code coverage for the following changes... can we override that github action?

![image](https://user-images.githubusercontent.com/43529054/103677665-94f51600-4f50-11eb-880c-7f22b95961e9.png)

![image](https://user-images.githubusercontent.com/43529054/103677688-9b838d80-4f50-11eb-830d-75848482d7d5.png)

I ran coverage locally for these files and nothing has changed between `master` and this branch as a sanity check.

pgrzesik commented on test/unit/lib/plugins/aws/deploy/lib/createStack.test.js:
```diff
@@ -126,11 +126,11 @@ describe('createStack', () => {
       sandbox.stub(awsDeploy, 'create').resolves();
       sandbox.stub(awsDeploy.provider, 'request').resolves();
-      awsDeploy.serverless.service.service = 'service-name'.repeat(100);
+      awsDeploy.serverless.service.provider.stackName = 'service-name'.repeat(100);
```
Is there a reason to change this test setup here?

ifitzsimmons replied:
Huh, I'm not sure how or why that snuck in there.... I did perform a rebase, but to answer your question - no, there's no reason that that should be there.

pgrzesik reviewed (changes requested):
Thank you @ifitzsimmons for the changes. I've inspected the coverage changes and it's possible that I gave you the wrong suggestion with dropping the `async` from `extendedValidate`. It is used in a few places and possibly it's expected from that function to return a Promise (it's used in `AwsInvoke` where we see the dropped coverage). Would you be able to change it back and validate again if we're experiencing coverage drop? Sorry for the trouble :bow:

ifitzsimmons commented:
> Thank you @ifitzsimmons for the changes. I've inspected the coverage changes and it's possible that I gave you the wrong suggestion with dropping the `async` from `extendedValidate`. It is used in a few places and possibly it's expected from that function to return a Promise (it's used in `AwsInvoke` where we see the dropped coverage). Would you be able to change it back and validate again if we're experiencing coverage drop? Sorry for the trouble 🙇

I think that maybe I caused that confusion. I removed all async/await logic associated with `extendedValidate()`. Before these changes, it returned `BBPromise.resolve()` without ever calling any functions that returned promises. I think your initial recommendation to remove async from extended validate was spot on.   Still having codecov issues since I removed the call to BBPromise.resolve. Adding a test case now to increase codecov

ifitzsimmons commented:
@pgrzesik Let me know if you want me to squash all of these commits before merging.

ifitzsimmons commented:
@pgrzesik Am I missing something here. When I click on the codecov report, I am seeing that extendedValidate() is the violator. But when I run codecoverage locally, my branch has more code coverage than when I run coverage locally on master.

My branch is on the left, master is on the right. 
![image](https://user-images.githubusercontent.com/43529054/103848374-3cb53580-5070-11eb-8f3a-c0ff2f0e432c.png)

pgrzesik commented:
I went through it once again and I believe it's just `coverage` playing with us here, it looks like it's all good, however, I think we should remove the last change with `{ image: true }` as it doesn't actually test anything extra as there are no new assertions in the test. Thanks for being patient with that PR :bow:

ifitzsimmons commented:
> I went through it once again and I believe it's just `coverage` playing with us here, it looks like it's all good, however, I think we should remove the last change with `{ image: true }` as it doesn't actually test anything extra as there are no new assertions in the test. Thanks for being patient with that PR 🙇

I'm happy to remove it, but it does add coverage. I added another function object, but no additional calls to `fileExistsSyncStub` were made because of the following line
```javascript
if (functionObject.image) return;
```

If that object did not have an `image` property, we'd have to change the line in the test case:
```diff
- expect(fileExistsSyncStub.calledTwice).to.equal(true);
+ expect(fileExistsSyncStub.calledThrice).to.equal(true);
```

Let me know if you still want me to remove that line!! I'm happy to do so

pgrzesik commented:
Hey @ifitzsimmons :wave: I understand that it might add coverage there, but the definition of that image is invalid (currently it expects a valid Docker Image URI) and there are no extra assertions added, so there's little value in added coverage if we don't really test the outcome of the tested path. I'll say let's remove the whole
```
second: {
          image: true,
        },
```
that was added, I don't think it will require changes in the assertions

ifitzsimmons commented:
@pgrzesik Are you waiting on me for anything for this PR?

pgrzesik commented:
Hello @ifitzsimmons, really sorry about it - I didn't realise that you pushed new changes because I'm monitoring review requests and comments via notifications and didn't get anything :( The best bet to ensure I or someone else sees the changes pushed is to request for re-review after completing changes.

Unfortunately, I see that there's one extra conflicting file that will need to be resolved before we can merge this one

ifitzsimmons commented:
@pgrzesik That codecov issue resurfaced, can we ignore it again?

pgrzesik reviewed (approved):
Awesome job :raised_hands: Coverage looks good, your diff has 100% of coverage :100:
